Use rayon to deserialize CFD events in parallel

thomaseizinger opened the pull request:
Rayon is a parallel data processing library. By replacing `into_iter`
with `into_par_iter`, we are constructing a parallel iterator that is
baked by a thread-pool which defaults to a number of threads equal to
the logical cores of the system.

On my local machine (8 logical cores), this speeds up the
`projection::Initialize` handler from 83 to 28 seconds for a 2.8 GB
database.

thomaseizinger committed:
Use rayon to deserialize CFD events in parallel

Rayon is a parallel data processing library. By replacing `into_iter`
with `into_par_iter`, we are constructing a parallel iterator that is
baked by a thread-pool which defaults to a number of threads equal to
the logical cores of the system.

On my local machine (8 logical cores), this speeds up the
`projection::Initialize` handler from 83 to 28 seconds for a 2.8 GB
database.

luckysori reviewed (approved):
Damn, that's amazing! Any downsides to using this?

klochowicz reviewed (approved):
awesome idea! 👍

da-kami reviewed (approved):
Great find, happy to give it a try!
Sounds promising! https://github.com/rayon-rs/rayon#no-data-races

thomaseizinger committed:
Ensure events are ordered chronologically

It is really important that we feed the events in chronological order to
the aggregate. Previously, this was done implicitly through the insertion
order in the DB.

Rayon will preserve ordering [0] of the original iterator but given how
important this ordering is for our logic to work correctly, it is better
to be explicit about it.

[0]: https://github.com/rayon-rs/rayon/issues/551

da-kami reviewed (approved):
I suppose time needed for sorting can be neglected.

bonomat commented on daemon/src/db.rs:
```diff
@@ -1123,6 +1125,8 @@ async fn load_cfd_events(
     })
     .collect::<Result<Vec<_>>>()?;
 
+    events.sort_unstable_by(CfdEvent::chronologically);
```
Why don't you use sql for the ordering? That would save us from needing `CfdEvent::chronologically`

thomaseizinger replied:
Because I want to ensure that I do the sorting _after_ I've used rayon to deserialize the events in parallel. I think rayon should preserve the order but as mentioned in the commit message, I think it is better to be explicit about this!

bonomat replied:
Ah, got it.

luckysori commented on model/src/cfd.rs:
````diff
@@ -378,6 +379,32 @@ impl CfdEvent {
+    /// ```rust
+    /// # use model::{CfdEvent, Timestamp, EventKind, OrderId};
+    ///
+    /// # let id = OrderId::default();
+    /// # let make_event_at = move |seconds| CfdEvent {
+    /// #    timestamp: Timestamp::new(seconds),
+    /// #    id,
+    /// #    event: EventKind::LockConfirmed,
+    /// # };
+    /// #
````
❓ Thanks for adding an example! What do the `#`s mean?

thomaseizinger replied:
They make this part of the code not render in rustdoc!
